source/_posts/2019-03-11-upgrading-to-embark-4.md

```diff
@@ -0,0 +1,73 @@
+title: How to upgrade to Embark 4
+summary: "In this guide, we'll show how to upgrade from an Embark 3.x project to Embark 4"
+categories:
+  - tutorials
+---
+
+The release of Embark 4.0 is close at hand and beta.2 is the relase to introduce some real breaking changes. Let's dive right into them.
+
+## Use the pipeline that YOU want
+
+You read correctly! You no longer need to use Embark's pipeline, which, let's not hide it, is pretty slow.
+
+Historically, Embark has come with a special Webpack pipeline because it made your work, as a user, easier. Indeed, it enabled you to use simple "magic" imports, such as `import SimpleStorage from "Embark/contracts/SimpleStorage";`. It also did the connection to Web3 for you.
+
+However, we discovered the hard way that those advantages ended not being worth the price of performance that you pay, when you can't use an optimized pipeline, like create-react-app or Angular CLI.
+
+Thereby, we announce that you can now use the file pipeline that YOU want. Also, you can STILL use Embark's pipeline if you are fine with what is currently there.
+
+So, as you just read, it's not really a breaking change, well  that is, unless you decide you want to convert you Dapp from Embark's pipeline to the one of your choice.
+
+If you don't plan on starting a project with another pipeline or don't plan of converting, you can skip to [the next section](#New-Web3-plugin).
+
+### Converting to another pipeline
+
+The first thing you might want to do is change or add the new `generationDir` property in `embark.json`. This property tells Embark where to generation the various artifacts that you will need to build your Dapp. Most of those artifacts were already generated before, but were inside the `.embark/` folder. 
+
+Since most file pipelines need those files inside the source folder, we're going to put `generationDir` as your selected pipeline's source folder. For example,  create-react-app (CRA) has `src/` as source folder.
+
+Afterwards, we need to convert all "magic" imports to relative imports.
+
+The first one is the EmbarkJS import. The "magic" import is `"Embark/EmbarkJS"`. We need to convert that to the relative path. The `embarkjs.js` script is at the root of  `embarkArtifacts/`, so you need to convert it to something like `"./embarkArtifacts/embarkjs"`.
```

Comment:
```suggestion
The first one is the EmbarkJS import. The "magic" import is `"Embark/EmbarkJS"`. Anywhere we have `"Embark/EmbarkJS"` in our Dapp's code, we need to convert that to the relative path. Because we are trying to get the `EmbarkJS` library, and the `embarkjs.js` script is located in the root of  `embarkArtifacts/`, we need to replace 
```
```
import EmbarkJS from "Embark/EmbarkJS"
```
with
```
import EmbarkJS from "./embarkArtifacts/embarkjs"
```
NOTE: The relative path is dependent upon the `generationDir` setting specified in `embark.json` (see the "Artifact generation directory" section above](#Artifact generation directory).